### PR TITLE
added datalock code for Functional Skills as well as auditing

### PR DESCRIPTION
**Pull request by davehowlett**

### CODE REVIEW

**Comment by carlbm** [comment]
The naming is a little confusing. There are Act2FunctionalSkills, PayableFunctionalSkills and FunctionalSkills. Not 100% clear what each represents? 
Also, there don't seem to be any non-payable functional skills. Are they included?

**Comment by davehowlett** [comment]
FunctionalSkillEarningEvent is a base event used by the EarningEvents classes Act1FunctionalSkillEarningEvent and Act2FunctionalSkillEarningEvent. @abroodie wanted them splitting so that there weren't any issues with NServicebus handling a base class in RequiredPayments for example.
PayableFunctionalSkillEarningEvent is what will come out of the DataLock component and comes from the Act1FunctionalSkillEarningEvent. In a similar vein to non payable earning events (which are generated using the EarningFailedDataLockMatching class), a FunctionalSkillEarningFailedDataLockMatching class has been created for non payable functional skills. There is nothing currently consuming this, just like EarningFailedDataLockMatching

File: src/SFA.DAS.Payments.Audit.Application/Mapping/EarningEvents/EarningEventProfile.cs [hunk 1]
```diff
@@ -29,10 +31,20 @@ public EarningEventProfile()
                 .ForMember(dest => dest.ContractType, opt => opt.UseValue(ContractType.Act2))
                 .ForMember(dest => dest.Periods,opt => opt.ResolveUsing<ApprenticeshipContractTypeEarningPeriodResolver>());
 
-            CreateMap<ApprenticeshipContractType2FunctionalSkillEarningsEvent, EarningEventModel>()
-                .ForMember(dest => dest.ContractType, opt => opt.UseValue(ContractType.Act2))  //TODO: fix for ACT1 events
+            CreateMap<Act1FunctionalSkillEarningsEvent, EarningEventModel>()
+                .ForMember(dest => dest.ContractType, opt => opt.UseValue(ContractType.Act1));
```
[the remark this answers is not in the document]
Comment (davehowlett) [reply src/SFA.DAS.Payments.Audit.Application/Mapping/EarningEvents/EarningEventProfile.cs]: there are two classes Act1FunctionalSkillMappingTests and Act2FunctionalSkillMappingTests that inherit from FunctionalSkillMappingTests<T> for Act1FunctionalSkillEarningEvent and Act2FunctionalSkillEarningEvent respectively. This base class does test these properties on both classes

File: src/SFA.DAS.Payments.Audit.Application/Mapping/EarningEvents/EarningEventProfile.cs [hunk 2]
```diff
@@ -29,10 +31,20 @@ public EarningEventProfile()
...
+            CreateMap<Act2FunctionalSkillEarningsEvent, EarningEventModel>()
+                .ForMember(dest => dest.ContractType, opt => opt.UseValue(ContractType.Act2));
+
+            CreateMap<FunctionalSkillEarningsEvent, EarningEventModel>()
+                .Include<Act1FunctionalSkillEarningsEvent, EarningEventModel>()
+                .Include<Act2FunctionalSkillEarningsEvent, EarningEventModel>()
                 .ForMember(dest => dest.Periods, opt => opt.ResolveUsing<FunctionalSkillEarningResolver>());
 
+            //CreateMap<Act2FunctionalSkillEarningsEvent, EarningEventModel>()
```
[the remark this answers is not in the document]
Comment (davehowlett) [reply src/SFA.DAS.Payments.Audit.Application/Mapping/EarningEvents/EarningEventProfile.cs]: good spot. No. Removed

File: src/SFA.DAS.Payments.Audit.Application.UnitTests/Mapping/EarningEvent/FunctionalSkillMappingTests.cs [hunk 3]
```diff
@@ -38,10 +37,10 @@ public void Maps_Earnings()
         }
 
 
-        [TestCaseSource(nameof(GetContractTypes))]
```
[the remark this answers is not in the document]
Comment (davehowlett) [reply src/SFA.DAS.Payments.Audit.Application.UnitTests/Mapping/EarningEvent/FunctionalSkillMappingTests.cs]: This class has been changed to be generic and the two classes for ACT1 and ACT2 handle what was in this test case source

**Pull request merged**